### PR TITLE
Prince/68609/POC - websocket hook

### PR DESCRIPTION
**Pull request by prince-deriv**
Changes:

-   ...

## Type of change

-   [ ] Bug fix
-   [ ] New feature
-   [ ] Update feature
-   [x] Refactor code
-   [ ] Translation to code
-   [ ] Translation to crowdin
-   [ ] Script configuration
-   [ ] Improve performance
-   [ ] Style only
-   [ ] Dependency update
-   [ ] Documentation update
-   [ ] Release

### CODE REVIEW

File: src/store/index.tsx [hunk 1]
```diff
@@ -63,3 +90,11 @@ export const DerivProvider = ({ children }: DerivProviderProps) => {
         </DerivStore.Provider>
     )
 }
+
+export const DerivSocket = () => {
```
Comment (balakrishna-binary) [inline src/store/index.tsx:94]:
```suggestion
export const useDerivSocket = () => {
```

Comment (balakrishna-binary) [reply src/store/index.tsx:94]: since this is a custom hook, we should be having `use` prefix

**Review by balakrishna-binary — DISMISSED** [review]

**Review by mohammad-rework — APPROVED** [review]

File: src/common/websocket/api.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    public reject(reason?: any) {
+        this.rejectFn(reason)
+    }
+
+    public resolve(param: T) {
+        this.resolveFn(param)
+    }
+}
+
+export default class DerivAPI {
```
Comment (habib-deriv) [inline src/common/websocket/api.ts:42]: Should we make this DerivWSApi, instead of DerivAPI?

File: src/common/websocket/api.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    private pendingRequests: {
+        [k: string]: CustomPromise<DerivAPIResponse>
+    } = {}
+    private reqId = 0
+    private connected = new CustomPromise<void>()
+    connection: WebSocket
+
+    constructor() {
+        const socket_url = `${getSocketURL()}?app_id=${getAppId()}&l=${
+            getLanguage() === 'ach' ? getCrowdin() : getLanguage()?.replace('-', '_')
```
Comment (habib-deriv) [inline src/common/websocket/api.ts:52]: Let's move this part outside and use this as a variable

File: src/common/websocket/api.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        const connection = new WebSocket(socket_url)
+        connection.onopen = this.openHandler
+        connection.onmessage = this.messageHandler
+
+        this.connection = connection
+    }
+
+    isConnectionClosed() {
+        return this.connection.readyState === 2 || this.connection.readyState === 3
```
Comment (habib-deriv) [inline src/common/websocket/api.ts:63]: `this.connection.readyState === 2`
For these numbers can we save them as a meaningfull variable name and use it to compare

**Review by mohammad-rework — APPROVED** [review]

**Review by habib-deriv — APPROVED** [review]

**Pull request merged**